src/paho/mqtt/client.py

```diff
@@ -523,11 +531,25 @@ def __init__(self, client_id="", clean_session=True, userdata=None,
         Set transport to "websockets" to use WebSockets as the transport
         mechanism. Set to "tcp" to use raw TCP, which is the default.
         """
-        if not clean_session and (client_id == "" or client_id is None):
-            raise ValueError('A client id must be provided if clean session is False.')
+
+        if protocol == MQTTv5:
+            if not clean_session:
```

Comment:
The names 'clean session' and 'clean start' are used in the MQTT specifications for 3.1.1 and 5.0, so I thought it was the right thing to do to use those names for the parameters.  

We could have one parameter called clean_session_or_start, but that would cause some backward compatibility problems.  Is that what you were thinking?

Reply:
I think of two options:

* Either keep clean_session and clean_start options but just change the default from True to None. This will also to raise the error as soon as user specify either True or False
* Or make clean_session an alias of clean_start... we may achieve this using **kwargs. But I fear this solution is too complicated for the gain.

Reply:
I agree that having clean_session and clean_start as synonyms would be ideal, but also complicated to implement.

Actually this brings up another consideration.  I was surprised to see the clean_session option on the constructor (I haven't used this Python client very much), rather than the connect method.

In MQTT 5.0, the session cleanup is governed by the session expiry property, which defaults to 0.  This means that if you want to clean up at the start of the session, but be able to reconnect and continue if there are communication breaks until your work is done, you can do it in one go:

connect clean_start=true, session_expiry=99999 (value in seconds)

disconnect also has a session expiry property defaulting to 0, so when you issue a normal disconnect, then the session is cleaned up immediately.

To me, it makes sense to have the clean_start and session_expiry property on the same method.  One way of making this work would be to have the clean_start parameter on connect rather than the constructor, along with the connect properties.  Then clean_session on the constructor would not conflict with clean_start, and would not be allowed with MQTT 5.0.